helpers/create_pnp_backdata.py

```diff
@@ -327,24 +350,77 @@ def populate_instance_1_columns(df):
     Return:
         df (pd.DataFrame): The dataframe with the populated instance 1 columns.
     """
-    df["period_reference"] = df["period"].astype(str) + \
-        df["reference"].astype(str) + \
-        df["instance"].astype(str)
-    df = df.set_index("period_reference")
-
-    source_df = df[df["instance"] == 0].copy()
-
-    for col in config["breakdowns"]["211"] + \
-        config["breakdowns"]["305"] + \
-            ["211", "305"]:
-        if col in source_df.columns:
-            for period_reference in source_df.index.values:
-                df.loc[period_reference[0:-1] + "1", col] = source_df.loc[
-                    period_reference, col
-                ]
-    df = df.reset_index()
-
-    return df
+    # a list of columns to be updated
+    numcols = config["breakdowns"]["211"] + config["breakdowns"]["305"] + ["211", "305"]
+    cols = [c for c in numcols if c in df.columns]
+
+    # the rows which contain the data use for the updates
+    source_df = df[df["instance"] == 0].copy()[["reference"]  + cols]
+    # the dataframe to be used for the update
+    update_df = source_df.copy()
+    update_df["instance"] = 1
+
+    # add extra rows with instance 1 to the original dataframe if a reference does not
+    # have an istance = 1 row
+    refs_with_ins_1 = df[df["instance"] == 1]["reference"].unique()
+    refs_without_ins_1 = set(source_df["reference"].unique()) - set(refs_with_ins_1)
+
+    extra_rows_df = df[df["reference"].isin(refs_without_ins_1)].copy()
+    extra_rows_df["instance"] = 1
+
+    df = pd.concat([df, extra_rows_df], ignore_index=True)
+
+    merged_df = pd.merge(df, update_df, on=["reference", "instance"], how="left", suffixes=("", "_y"))

```

Comment:
Can we please have this line over several, to not exceed the character limit per line. Cheers.